checkrelease.sh script now tries to build sim:nsh

adamfeuer opened the pull request:
## Summary

- checkrelease.sh script now builds sim:nsh
- usage has changed to use command line options
- can check a directory on disk as well as a URL or release name

## Impact

- none if you don't use the script
- script usage options have changed

## Usage

```
Usage: /home/adam/src/nuttx/nuttx-sama5d3-xplained/nuttx/tools/checkrelease.sh [--verbose] [--gpg-output] [--url <URL-of-release-dir>] [--release <name-of-release] [--dir <path-to-directory>] [--tempdir <path-to-directory>]
   Given release full URL, release name, or a local directory, downloads or copies
   all files in that directory (which for a release should include nuttx and nuttx-apps, sha512, 
   asc, and tar.gz files), checks the release SHA512 and GPG signatures, checks the unpacked 
   directories for required files, and tries to build NuttX for sim:nsh.

   If tempdir is specified, it will be removed and recreated; if it is not specified, /tmp/nuttx-checkrelease
   is used.

   If --url or --release are given, nuttx and nuttx-apps tar.gz files are left in /tmp for the user to 
   build for their platform.

   -V and --verbose are equivalent.
   -G and --gpg-output are equivalent.

Examples:

  ./checkrelease.sh --release 9.1.0-RC1
  ./checkrelease.sh --url https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
  ./checkrelease.sh --dir ./some-dir-that-has-nuttx-and-apps
```

## Testing

- manual, I used 9.1.0-RC1 and a modified working version of 9.1.0-RC1 (with `init.d/rcS` file)

adamfeuer commented:
@xiaoxiang781216 I added the `sim:nsh` build to the `checkrelease.sh` script as you suggested. Will you take a look?

btashton commented:
@adamfeuer I gave it ago against the release I just staged and it seemed to work.  A couple comments:
1. When I put a bad release name in `9.1.0-RC2A` the downloads 404ed but the script kept running all the checks.  We should probably make sure we have the files that we are looking for.
2. Can we quiet the wget it

Thanks for putting this together.  The dir option is nice for checking prior to committing to the release staging SVN.

```
❯ curl -s https://raw.githubusercontent.com/apache/incubator-nuttx/9bc97e7c0dfe0d392e39f11419081cf48db3b8fd/tools/checkrelease.sh | bash /dev/stdin --release 9.1.0-RC2
--2020-07-07 17:43:51--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/
Resolving dist.apache.org (dist.apache.org)... 13.90.137.153
Connecting to dist.apache.org (dist.apache.org)|13.90.137.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 886 [text/html]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/index.html’

dist.apache.org/index.html                100%[==================================================================================>]     886  --.-KB/s    in 0s      

2020-07-07 17:43:59 (12.3 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/index.html’ saved [886/886]

Loading robots.txt; please ignore errors.
--2020-07-07 17:43:59--  https://dist.apache.org/robots.txt
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 165 [text/plain]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/robots.txt’

dist.apache.org/robots.txt                100%[==================================================================================>]     165  --.-KB/s    in 0s      

2020-07-07 17:44:00 (180 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/robots.txt’ saved [165/165]

--2020-07-07 17:44:00--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-9.1.0-incubating.tar.gz
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 23767677 (23M) [application/gzip]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz’

dist.apache.org/apache-nuttx-9.1.0-incuba 100%[==================================================================================>]  22.67M  8.90MB/s    in 2.5s    

2020-07-07 17:44:05 (8.90 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz’ saved [23767677/23767677]

--2020-07-07 17:44:05--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-9.1.0-incubating.tar.gz.asc
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 833 [application/pgp-signature]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz.asc’

dist.apache.org/apache-nuttx-9.1.0-incuba 100%[==================================================================================>]     833  --.-KB/s    in 0s      

2020-07-07 17:44:07 (34.1 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz.asc’ saved [833/833]

--2020-07-07 17:44:07--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-9.1.0-incubating.tar.gz.sha512
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 167 [text/plain]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz.sha512’

dist.apache.org/apache-nuttx-9.1.0-incuba 100%[==================================================================================>]     167  --.-KB/s    in 0s      

2020-07-07 17:44:09 (2.99 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-9.1.0-incubating.tar.gz.sha512’ saved [167/167]

--2020-07-07 17:44:09--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-apps-9.1.0-incubating.tar.gz
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 3756073 (3.6M) [application/gzip]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz’

dist.apache.org/apache-nuttx-apps-9.1.0-i 100%[==================================================================================>]   3.58M  5.89MB/s    in 0.6s    

2020-07-07 17:44:11 (5.89 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz’ saved [3756073/3756073]

--2020-07-07 17:44:11--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 833 [application/pgp-signature]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz.asc’

dist.apache.org/apache-nuttx-apps-9.1.0-i 100%[==================================================================================>]     833  --.-KB/s    in 0s      

2020-07-07 17:44:14 (13.3 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz.asc’ saved [833/833]

--2020-07-07 17:44:14--  https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC2/apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
Reusing existing connection to dist.apache.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 172 [text/plain]
Saving to: ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512’

dist.apache.org/apache-nuttx-apps-9.1.0-i 100%[==================================================================================>]     172  --.-KB/s    in 0s      

2020-07-07 17:44:16 (2.81 MB/s) - ‘/tmp/nuttx-checkrelease/dist.apache.org/apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512’ saved [172/172]

FINISHED --2020-07-07 17:44:16--
Total wall clock time: 25s
Downloaded: 8 files, 26M in 3.2s (8.32 MB/s)
Checking apache-nuttx-9.1.0-incubating.tar.gz sha512...
 OK: apache-nuttx-9.1.0-incubating.tar.gz sha512 hash matches.

Checking apache-nuttx-9.1.0-incubating.tar.gz GPG signature:
 OK: apache-nuttx-9.1.0-incubating.tar.gz gpg signature matches.

Checking apache-nuttx-9.1.0-incubating.tar.gz for required files:
 OK: all required files exist in nuttx.

Checking apache-nuttx-apps-9.1.0-incubating.tar.gz sha512...
 OK: apache-nuttx-apps-9.1.0-incubating.tar.gz sha512 hash matches.

Checking apache-nuttx-apps-9.1.0-incubating.tar.gz GPG signature:
 OK: apache-nuttx-apps-9.1.0-incubating.tar.gz gpg signature matches.

Checking apache-nuttx-apps-9.1.0-incubating.tar.gz for required files:
 OK: all required files exist in apps.

Trying to build nuttx sim:nsh...
 OK: we were able to build sim:nsh
```

adamfeuer commented:
@btashton I will quiet the `wget`. Re: not finding the file... should we quit immediately then...?

adamfeuer commented:
@btashton I pushed a change that checks the URL, if it doesn't exist the script quits immediately. I also quieted the wget.

adamfeuer commented:
@btashton I'll see if I can add the GPG output back using the `--verbose` flag. Would that work?

adamfeuer commented:
@btashton I added a `-G` and `--gpg-output` flag to the script. If you give that option, it will show the verbose GPG output. Let me know if this works for you.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+    rm -rf "$RELEASE_DIR"
+    tar xf "$RELEASE_FILE"
+    check_required_files "$RELEASE_DIR"
+    mv "$RELEASE_FILE" "$TMP"
```
Why not reuse TEMPDIR?

adamfeuer replied:
@xiaoxiang781216 I'll try that out and see if it works.

adamfeuer replied:
@xiaoxiang781216 This is moving the nuttx and nuttx-apps to the /tmp dir so they are there for the user to unpack and build. This the reason I separated `$TMP` and `$TEMPDIR`. I'm going to leave them in for now. If you have a good way to leave the files somewhere like /tmp, let me know. I can't use `$TEMPDIR` because the trap removes it.

xiaoxiang781216 replied:
If so, why we don't remove trap directly to keep the result?

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+    echo "Trying to build $RELEASE_DIR sim:nsh..."
+    cd "$RELEASE_DIR"
+    if [[ $VERBOSE -eq 0 ]]; then
+      output=$(make distclean ; ./tools/configure.sh sim:nsh ; make) 2>&1
```
Move 2>&1 into ()

adamfeuer replied:
@xiaoxiang781216 If I do that it, I think it will only apply to the last command, not all of them. I don't think that's what you want— is it?

xiaoxiang781216 replied:
If so, how about we change (2>&1) out $(gpg ...) too?

adamfeuer replied:
@xiaoxiang781216 I thought more about @btashton's comment above, and removed the option to quiet the GPG invocation completely. He suggested seeing the output since there's info there that would be lost with a simple "Ok". So I'm having the GPG output show on every run.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+if [[ -n "$URL" ]]; then
+  check_nuttx
+  check_nuttx_apps
+else
```
Why not directly call check_nuttx and check_nuttx_apps?

adamfeuer replied:
@xiaoxiang781216 I'm not sure what you mean here— do you mean, why am I checking for the $URL? If so, it's because there's an option to check a directory on disk, and those functions assume a URL is given... and so will try to download files...

xiaoxiang781216 replied:
But check_nuttx/check_nuttx_apps should work with local directory too if my understanding is correct.

adamfeuer replied:
@xiaoxiang781216 The check_nuttx/check_nuttx_apps expect tar.gz files and try to decompress them. I thought about adding arguments to suppress them, but factoring the code this way seemed cleaner.

xiaoxiang781216 replied:
I guess what @btashton want is check the release is good before uploading to https://dist.apache.org/repos/dist/dev/incubator/nuttx/. So the local directory should contain the same files as the website.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
-    usage
-    exit 0
-fi
+POSITIONAL=()
```
change to UNKNOWN

adamfeuer replied:
@xiaoxiang781216 will fix

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+    ;;
+    *)    # unknown option
+    POSITIONAL+=("$1") # save it in an array for later
+    shift # past argument
```
Move shift after esac but before done, so we can avoid dup shift in each case.

adamfeuer replied:
@xiaoxiang781216 I will look at this to see if I can do it.

adamfeuer replied:
@xiaoxiang781216 fixed

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -28,13 +28,23 @@ TEMPDIR="$TMP/nuttx-checkrelease"
 ORIGINAL_DIR="$(pwd)"
 trap "{ cd $ORIGINAL_DIR; rm -rf $TEMPDIR; }" EXIT
 
+function validate_url(){
```
add space before {

adamfeuer replied:
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+TMP="/tmp"
+TEMPDIR="$TMP/nuttx-checkrelease"
 ORIGINAL_DIR="$(pwd)"
 trap "{ cd $ORIGINAL_DIR; rm -rf $TEMPDIR; }" EXIT
```
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
-TEMPDIR="dist.apache.org"
+DIST_DIR="dist.apache.org"
+TMP="/tmp"
+TEMPDIR="$TMP/nuttx-checkrelease"
```
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+      echo "One of --dir or --url is required!"
+      exit 1
+    fi
+  fi
```
@xiaoxiang781216 I'll try this out to see how it looks—

adamfeuer replied:
@xiaoxiang781216 I think I want to leave the code as is. I tried the `wget` command options you suggest, but they don't work exactly the same way that @btashton's do. Are you ok if we leave as-is?

xiaoxiang781216 replied:
Here is my output:
```
wget -q -r -nd -np https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
tree
.
├── apache-nuttx-9.1.0-incubating.tar.gz
├── apache-nuttx-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512
├── apache-nuttx-apps-9.1.0-incubating.tar.gz
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
├── index.html
└── robots.txt
```
What's different from yours?

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
-    tar xf "$RELEASE_FILE"
+    RELEASE_DIR=$1
+    echo "Checking $RELEASE_FILE for required files:"
     ERROR=0
```
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
     RELEASE_FILE=$1
     echo "Checking $RELEASE_FILE sha512..."
-    sha512sum -c "$RELEASE_FILE.sha512"
+    output="$(sha512sum -c $RELEASE_FILE.sha512)"
```
@xiaoxiang781216 I'll make these consistent

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
     check_gpg "$RELEASE_FILE"
-    check_required_files "$RELEASE_FILE" "nuttx"
-    mv "$RELEASE_FILE" ..
+    rm -rf "$RELEASE_DIR"
```
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
     check_gpg "$RELEASE_FILE"
-    check_required_files "$RELEASE_FILE" "apps"
-    mv "$RELEASE_FILE" ..
+    rm -rf "$RELEASE_DIR"
```
@xiaoxiang781216 will fix

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
+    RELEASE_DIR="nuttx"
+    echo "Trying to build $RELEASE_DIR sim:nsh..."
+    cd "$RELEASE_DIR"
+    if [[ $VERBOSE -eq 0 ]]; then
```
@xiaoxiang781216 The problem with that is if there is a failure, the error text is often quite long, and obscures the other output. I'll see what I can do to solve that problem.

adamfeuer replied:
@xiaoxiang781216 I tried it the way you suggest— what happens is all the newlines get stripped out when being captured, so the output is unreadable for long things like the nuttx compilation. The `--verbose` option lets you see the compilation formatted like you normally see it, it's a lot more readable. I'd like to leave the verbose switch in. But if you feel strongly about it, I'll do it your way, let me know.

xiaoxiang781216 replied:
Add quoto like this:
```
echo "$output"
```
You will get the corret format. https://stackoverflow.com/questions/4651437/how-do-i-set-a-variable-to-the-output-of-a-command-in-bash

adamfeuer commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,218 @@
     echo
-    echo "   nuttx and nuttx-apps tar.gz files are left in the current"
-    echo "   directory."
+    echo "   If --url or --release are given, nuttx and nuttx-apps tar.gz files are left in /tmp for the user to "
```
@xiaoxiang781216 will fix

adamfeuer commented:
@btashton @xiaoxiang781216 Ok, pushed changes that incorporate most of @xiaoxiang781216's suggestions.

What do you think?

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,221 @@
-    exit 0
-fi
+  URL=""
+  DIRECTORY=""
```
We clear URL/DIRECTORY in each iteration?

adamfeuer replied:
@xiaoxiang781216 that's a mistake, will fix

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,221 @@
-  URL="$1/"
-else
-  URL="$BASE_URL/$1/"
+  case $key in
```
key is used only once, why not directly use "case $1" here?

adamfeuer replied:
@xiaoxiang781216 using code as documentation, for clarity. I will take this out if you want, let me know.

xiaoxiang781216 replied:
I am fine.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,221 @@
+      ;;
+      *)    # unknown option
+      POSITIONAL+=("$1") # save it in an array for later
+      shift # past argument
```
Remove, otherwise we lose the half unknow argument.

adamfeuer replied:
@xiaoxiang781216 will remove

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,221 @@
+  case $key in
+      -U|--url)
+      URL="$2"
+      shift
```
It's better change all $2 access to:
```
shift
URL=$1
```
so we always access variable by $1.

adamfeuer replied:
@xiaoxiang781216 will fix

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,221 @@
+      RETURN_CODE=1
+      echo " - Error checking $RELEASE_FILE gpg signature:"
+      echo "$output"
+      echo
```
redundant?

adamfeuer replied:
@xiaoxiang781216 The extra `echo` is there for formatting, to leave a blank line. I can take it out if you want, let me know.

adamfeuer replied:
@xiaoxiang781216 re: wget output, here's mine:

```
$ tree
.

0 directories, 0 files
$ wget -q -r -nd -np https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1
$ tree
.
├── 9.1.0-RC1
├── apache-nuttx-9.0.0-incubating.tar.gz
├── apache-nuttx-9.0.0-incubating.tar.gz.1
├── apache-nuttx-9.0.0-incubating.tar.gz.asc
├── apache-nuttx-9.0.0-incubating.tar.gz.asc.1
├── apache-nuttx-9.0.0-incubating.tar.gz.sha512
├── apache-nuttx-9.0.0-incubating.tar.gz.sha512.1
├── apache-nuttx-9.1.0-incubating.tar.gz
├── apache-nuttx-9.1.0-incubating.tar.gz.1
├── apache-nuttx-9.1.0-incubating.tar.gz.2
├── apache-nuttx-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-9.1.0-incubating.tar.gz.asc.1
├── apache-nuttx-9.1.0-incubating.tar.gz.asc.2
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512.1
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512.2
├── apache-nuttx-apps-9.0.0-incubating.tar.gz
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.1
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.asc
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.asc.1
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.sha512
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.sha512.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.2
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc.2
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512.2
├── index.html
├── index.html.1
├── index.html.2
├── index.html.3
├── index.html.4
├── KEYS
└── robots.txt

0 directories, 38 files
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```

On my system that invocation downloads extra files. You can see from the first `tree` command above I'm starting with an empty directory.

xiaoxiang781216 replied:
I hit the similar issue only when the temp directory isn't empty and wget rename the new download to *.1.

adamfeuer replied:
@xiaoxiang781216 I've tried your invocation several times in empty directories, and get the same result as above. What temp directory are you referring to?

adamfeuer replied:
@xiaoxiang781216 Here's the output from `wget --version`:

```
GNU Wget 1.19.4 built on linux-gnu.

-cares +digest -gpgme +https +ipv6 +iri +large-file -metalink +nls 
+ntlm +opie +psl +ssl/openssl 

Wgetrc: 
    /etc/wgetrc (system)
Locale: 
    /usr/share/locale 
Compile: 
    gcc -DHAVE_CONFIG_H -DSYSTEM_WGETRC="/etc/wgetrc" 
    -DLOCALEDIR="/usr/share/locale" -I. -I../../src -I../lib 
    -I../../lib -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_LIBSSL -DNDEBUG 
    -g -O2 -fdebug-prefix-map=/build/wget-Xb5Z7Y/wget-1.19.4=. 
    -fstack-protector-strong -Wformat -Werror=format-security 
    -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall 
Link: 
    gcc -DHAVE_LIBSSL -DNDEBUG -g -O2 
    -fdebug-prefix-map=/build/wget-Xb5Z7Y/wget-1.19.4=. 
    -fstack-protector-strong -Wformat -Werror=format-security 
    -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall -Wl,-Bsymbolic-functions 
    -Wl,-z,relro -Wl,-z,now -lpcre -luuid -lidn2 -lssl -lcrypto -lpsl 
    ftp-opie.o openssl.o http-ntlm.o ../lib/libgnu.a 

Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later
<http://www.gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Originally written by Hrvoje Niksic <hniksic@xemacs.org>.
Please send bug reports and questions to <bug-wget@gnu.org>.
```

What's your wget version?

adamfeuer replied:
@xiaoxiang781216 aha. I figured it out. You need the trailing slash at the end of the URL. If I place it there, no duplicate files are created. I don't know why. I'll change the code to use your invocation and ensure there is a trailing slash at the end of the URL.

adamfeuer commented:
@xiaoxiang781216 Ok, I pushed changes to address the resolved comments. I have questions above— the wget invocation you suggest leaves extra files in my directory. I'm on Ubuntu Linux 18.02. See above for the actual output The other unresolved stuff seems stylistic, I'll do it your way if you feel strongly about it, just let me know. 

Thanks for being willing to look at this code! :)

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,225 @@
+  check_required_files "nuttx"
+  check_required_files "apps"
+fi
+check_sim_nsh_build
```
Do you try build sim:nsh with the download from website? I don't think the script can work as expect, because:
1.the unzip nuttx/apps is moved to /tmp
2.but we try to build /tmp/nuttx-release/nuttx

adamfeuer replied:
@xiaoxiang781216 I'm not sure what you mean here. Yes, I'm testing the script several times with each modification. You can comment out the trap statement, add `--verbose`, run the script, and verify that `sim:nsh` is built in the correct directory. When I do that I can see it's using the correct temporary directory `/tmp/nuttx-checkrelease`.

adamfeuer replied:
@xiaoxiang781216 Ok, you might be right, something is messed up with how I'm testing. Looking into it.

adamfeuer replied:
@xiaoxiang781216 Ah, it's a difference in your invocation of wget vs. mine. When I change to yours, the dist.apache.org directory is not created. So I was testing correctly before, but when I changed to your wget just now, that broke the script.

adamfeuer replied:
@xiaoxiang781216 Ok, pushed changes to make the `wget` invocation work the way you suggested. It requires a trailing slash at the end of every URL. If the user gives a URL I add the slash because extra slashes are ok.

My testing setup was ok— all the options (`--url`, `--release`, `--dir`) all work for me.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,225 @@
-    usage
-    exit 0
+if [[ (-n "$UNKNOWN") ]]; then
+  usage
```
If we don't support the unknown argument, why we don't directly quit at line 215:
```
* )
usage
exit 1
```

adamfeuer replied:
@xiaoxiang781216 Yes, good idea, will remove the UNKNOWN processing.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,225 @@
+    else
+      RETURN_CODE=1
+      echo " - Error checking $RELEASE_FILE gpg signature:"
+      echo "$output"
```
If you don't capture the output, why not remove `echo "$output"`

adamfeuer replied:
@xiaoxiang781216 Yes, that's a mistake, will remove.

adamfeuer commented:
@xiaoxiang781216 Ok, I think I implemented all your suggestions. Anything else left to do?

xiaoxiang781216 commented:
It's almost done, I squanch your change and make some clean:
https://github.com/xiaoxiang781216/incubator-nuttx/commit/b2fd53ae0bdae78ad73b4ae722756789d05a4ab7
Other major change is:
1.Keep all content in /tmp/nuttx-release
2.Assume the content in local folder same as website
The first change help we identify the root cause if the check fail.
The second change help we find the problem just before the release is uploaded to the website.

adamfeuer commented:
@xiaoxiang781216 Your changes look good and I cherry-picked them, but they break the `--dir` option for me. I also pushed a small change to fix that `--dir` option problem. If you're ok with this, please merge; or if you want me to squash and rebase against master I can do that too, let me know. Or if you want more changes, let me know.

xiaoxiang781216 commented on tools/checkrelease.sh:
```diff
@@ -19,114 +19,208 @@
+    echo
 }
 
+function check_sim_nsh_build() {
```
Duplicate with line 124, remove?
Let's squash the commit to make the history clear. Other change looks good.

adamfeuer replied:
@xiaoxiang781216 Thanks for catching — that's the result of a bad merge. Will fix. I'll squash and force push after that.

adamfeuer commented:
@xiaoxiang781216 Ok, I removed the duplicate method from the bad merge, retested all the options, squashed, and force pushed. Will you look and see if this is ok? And merge if it is?

xiaoxiang781216 commented:
LGTM. Since the build check can't  verify the new script, let's merge it now.

xiaoxiang781216 commented:
@btashton the remaining feature is the sign verification, how to find the correct public key for gpg?

btashton commented:
You can do 
```
curl https://dist.apache.org/repos/dist/dev/incubator/nuttx/KEYS | gpg --import
```

That will import any public keys used by the project for signing. Currently it just contains mine but others would be appended if used.

btashton commented:
But this does do the check already. So maybe I don't understand what you are asking for?

adamfeuer commented:
@xiaoxiang781216 @btashton Thank you for the feedback and help with this script. It's much better code now. 🙂 👍

xiaoxiang781216 commented:
> You can do
> 
> ```
> curl https://dist.apache.org/repos/dist/dev/incubator/nuttx/KEYS | gpg --import
> ```
> 

So the user have to import KEYS manually.

> That will import any public keys used by the project for signing. Currently it just contains mine but others would be appended if used.

> But this does do the check already. So maybe I don't understand what you are asking for?

Yes, the script will call gpg, but gpg complain the sign mismatch if user doesn't import KEY yet(most people don't know how to do this). Does it make sense that checkrelease.sh download KEY files too and pass them to gpg directly?

btashton commented:
> > You can do
> > 
> > ```
> > curl https://dist.apache.org/repos/dist/dev/incubator/nuttx/KEYS | gpg --import
> > ```
> > 
> 
> So the user have to import KEYS manually.
They are also published in a public keys server, but yeah we don't know what keyid will be used so for updating pulling the latest KEYS file might be best. 

> 
> > That will import any public keys used by the project for signing. Currently it just contains mine but others would be appended if used.
> 
> > But this does do the check already. So maybe I don't understand what you are asking for?
> 
> Yes, the script will call gpg, but gpg complain the sign mismatch if user doesn't import KEY yet(most people don't know how to do this). Does it make sense that checkrelease.sh download KEY files too and pass them to gpg directly?

I only worry about messing with someone's local keystore without being explict but maybe its not that big of a worry since our how-to check also suggests this. 

If we wanted to use a temporary keystore (which would also make sure to only check against the KEYS file) we could do
```
export GNUPGHOME="$tmpstore"
gpg --import KEYS
...
```
